src/mongocxx/exception/base.hpp

```diff
@@ -16,13 +16,41 @@
 
 #include <mongocxx/config/prelude.hpp>
 
+#include <cstdint>
 #include <exception>
+#include <tuple>
+
+#include <bsoncxx/document/value.hpp>
+#include <bsoncxx/document/view.hpp>
+#include <bsoncxx/stdx/optional.hpp>
+
+#include <mongocxx/stdx.hpp>
 
 namespace mongocxx {
 MONGOCXX_INLINE_NAMESPACE_BEGIN
 namespace exception {
 
-class MONGOCXX_API base : public std::exception {};
+using error_and_code_t = std::tuple<std::string, std::int32_t>;
+
+class MONGOCXX_API base : public virtual std::exception {
+   public:
+    base(bsoncxx::document::value raw_server_error);
+    base(error_and_code_t error_and_code);
+    base(bsoncxx::document::value raw_server_error, error_and_code_t error_and_code);
+
+    ///
+    /// @returns The raw server error, if it is available.
+    ///
+    const stdx::optional<bsoncxx::document::value>& raw_server_error();
```

Comment:
Should there be non-const overloads, so that you could move from the optionals if you wanted?